File: shore-daemon/src/memory/deferred_edits.rs
use std::collections::BTreeSet;
use std::fs;
use std::io::{self, Write};
use std::path::{Path, PathBuf};

use shore_config::{
    character_config_dir, character_memory_dir, character_workspace_dir, character_workspace_file,
    AGENTS_FILE, HEARTBEAT_FILE, SOUL_FILE, TOOLS_FILE, USER_FILE,
};

/// Top-level workspace files that are editable immediately but only become
/// prompt-active after the next compaction/reload boundary.
const PROTECTED_PATHS: &[&str] = &[
    SOUL_FILE,
    USER_FILE,
    AGENTS_FILE,
    TOOLS_FILE,
    HEARTBEAT_FILE,
];

/// Persisted snapshot directory under the character data dir.
const ACTIVE_PROMPT_DIR: &str = "active_prompt";

/// Compaction-produced recent-memory digest injected into normal turns.
pub const RECENT_MEMORY_DIGEST_FILE: &str = "RECENT_MEMORY.md";

const DEFAULT_TOOLS_GUIDANCE: &str = "\
# TOOLS

Use tools when they materially help.

- Read files before editing them.
- Search memory files before guessing facts about the user or past events.
- Prefer concise, direct tool use over busywork.
";

const DEFAULT_HEARTBEAT_GUIDANCE: &str = "\
# HEARTBEAT

- Check whether anything from recent memory files needs follow-up.
- Check whether you promised the user anything that still matters.
- If nothing needs action, respond HEARTBEAT_OK.
";

/// Normalize a workspace path to one of the protected workspace-root files.
pub fn normalize_protected_path(path: &str) -> Option<String> {
    let mut normalized = path
        .trim()
        .trim_start_matches('/')
        .trim_start_matches('\\')
        .replace('\\', "/");

    while let Some(rest) = normalized.strip_prefix("./") {
        normalized = rest.to_string();
    }

    if let Some(rest) = normalized.strip_prefix("workspace/") {
        normalized = rest.to_string();
    }

    PROTECTED_PATHS
        .iter()
        .any(|&protected| normalized == protected)
        .then_some(normalized)
}

pub fn is_protected_path(path: &str) -> bool {
    normalize_protected_path(path).is_some()
}

pub fn active_prompt_dir(character_data_dir: &Path) -> PathBuf {
    character_data_dir.join(ACTIVE_PROMPT_DIR)
}

pub fn active_prompt_file(character_data_dir: &Path, name: &str) -> PathBuf {
    active_prompt_dir(character_data_dir).join(name)
}

pub fn recent_memory_digest_path(character_data_dir: &Path) -> PathBuf {
    active_prompt_file(character_data_dir, RECENT_MEMORY_DIGEST_FILE)
}

pub fn load_active_prompt_file(character_data_dir: &Path, name: &str) -> Option<String> {
    fs::read_to_string(active_prompt_file(character_data_dir, name))
        .ok()
        .filter(|content| !content.trim().is_empty())
}

pub fn write_recent_memory_digest(character_data_dir: &Path, digest: &str) -> io::Result<()> {
    let path = recent_memory_digest_path(character_data_dir);
    if let Some(parent) = path.parent() {
        fs::create_dir_all(parent)?;
    }
    fs::write(path, digest.trim().to_owned() + "\n")
}

/// Return deduped protected paths waiting for activation.
pub fn pending_deferred_edit_paths(character_data_dir: &Path) -> io::Result<Vec<String>> {
    let queue_path = character_data_dir.join("deferred_edits.jsonl");
    if !queue_path.exists() {
        return Ok(vec![]);
    }

    let content = fs::read_to_string(queue_path)?;
    let mut paths = BTreeSet::new();
    for line in content.lines() {
        let line = line.trim();
        if line.is_empty() {
            continue;
        }
        let entry: serde_json::Value = match serde_json::from_str(line) {
            Ok(v) => v,
            Err(_) => continue,
        };
        if let Some(path) = entry
            .get("path")
            .and_then(|v| v.as_str())
            .and_then(normalize_protected_path)
        {
            paths.insert(path);
        }
    }
    Ok(paths.into_iter().collect())
}

/// Queue a deferred refresh for a protected file.
pub fn queue_deferred_edit(character_data_dir: &Path, path: &str) -> io::Result<()> {
    let Some(path) = normalize_protected_path(path) else {
        return Ok(());
    };

    fs::create_dir_all(character_data_dir)?;
    let queue_path = character_data_dir.join("deferred_edits.jsonl");
    let mut file = fs::OpenOptions::new()
        .create(true)
        .append(true)
        .open(&queue_path)?;
    let line = serde_json::json!({
        "path": path,
        "timestamp": chrono::Local::now().to_rfc3339(),
    });
    writeln!(file, "{line}")?;
    Ok(())
}

/// Refresh the active prompt snapshot from the canonical config workspace and
/// clear any deferred-edit queue.
pub fn apply_deferred_edits(
    character_data_dir: &Path,
    config_dir: &Path,
    char_name: &str,
) -> io::Result<()> {
    refresh_active_prompt_snapshot(character_data_dir, config_dir, char_name)?;

    let queue_path = character_data_dir.join("deferred_edits.jsonl");
    if queue_path.exists() {
        fs::remove_file(queue_path)?;
    }
    Ok(())
}

/// Ensure the workspace-first layout exists for a character and migrate any
/// legacy bootstrap files or memory files into it.
pub fn ensure_character_workspace(
    character_data_dir: &Path,
    config_dir: &Path,
    char_name: &str,
) -> io::Result<()> {
    let char_config_dir = character_config_dir(config_dir, char_name);
    let workspace_dir = character_workspace_dir(config_dir, char_name);
    let memory_dir = character_memory_dir(config_dir, char_name);

    fs::create_dir_all(&workspace_dir)?;
    fs::create_dir_all(&memory_dir)?;

    migrate_legacy_file(
        char_config_dir.join("character.md"),
        workspace_dir.join(SOUL_FILE),
    )?;
    migrate_legacy_file(
        char_config_dir.join("user.md"),
        workspace_dir.join(USER_FILE),
    )?;
    migrate_legacy_file(
        char_config_dir.join("prompts").join("system.md"),
        workspace_dir.join(AGENTS_FILE),
    )?;

    let global_user = config_dir.join("user.md");
    let workspace_user = workspace_dir.join(USER_FILE);
    if global_user.exists() && !workspace_user.exists() {
        fs::copy(global_user, workspace_user)?;
    }

    write_default_if_missing(workspace_dir.join(TOOLS_FILE), DEFAULT_TOOLS_GUIDANCE)?;
    write_default_if_missing(
        workspace_dir.join(HEARTBEAT_FILE),
        DEFAULT_HEARTBEAT_GUIDANCE,
    )?;

    let legacy_memories = character_data_dir.join("memories");
    if legacy_memories.exists() {
        copy_tree_if_missing(&legacy_memories, &memory_dir)?;
    }

    Ok(())
}

/// Ensure the active prompt snapshot exists. Missing files are seeded from the
/// canonical config workspace, but an existing snapshot is left intact so edits
/// remain deferred until compaction.
pub fn ensure_active_prompt_snapshot(
    character_data_dir: &Path,
    config_dir: &Path,
    char_name: &str,
) -> io::Result<()> {
    ensure_character_workspace(character_data_dir, config_dir, char_name)?;

    let active_dir = active_prompt_dir(character_data_dir);
    fs::create_dir_all(&active_dir)?;

    for name in PROTECTED_PATHS {
        let dst = active_dir.join(name);
        if !dst.exists() {
            let src = character_workspace_file(config_dir, char_name, name);
            if src.exists() {
                fs::copy(src, dst)?;
            }
        }
    }

    Ok(())
}

/// Refresh every protected file in the active prompt snapshot.
pub fn refresh_active_prompt_snapshot(
    character_data_dir: &Path,
    config_dir: &Path,
    char_name: &str,
) -> io::Result<()> {
    ensure_character_workspace(character_data_dir, config_dir, char_name)?;

    let active_dir = active_prompt_dir(character_data_dir);
    fs::create_dir_all(&active_dir)?;

    for name in PROTECTED_PATHS {
        let src = character_workspace_file(config_dir, char_name, name);
        let dst = active_dir.join(name);

        if src.exists() {
            fs::copy(src, dst)?;
        } else if dst.exists() {
            fs::remove_file(dst)?;
        }
    }

    Ok(())
}

fn migrate_legacy_file(src: PathBuf, dst: PathBuf) -> io::Result<()> {
    if src.exists() && !dst.exists() {
        if let Some(parent) = dst.parent() {
            fs::create_dir_all(parent)?;
        }
        fs::copy(src, dst)?;
    }
    Ok(())
}

fn write_default_if_missing(path: PathBuf, content: &str) -> io::Result<()> {
    if path.exists() {
        return Ok(());
    }
    if let Some(parent) = path.parent() {
        fs::create_dir_all(parent)?;
    }
    fs::write(path, content)
}

fn copy_tree_if_missing(src: &Path, dst: &Path) -> io::Result<()> {
    if src.is_file() {
        if !dst.exists() {
            if let Some(parent) = dst.parent() {
                fs::create_dir_all(parent)?;
            }
            fs::copy(src, dst)?;
        }
        return Ok(());
    }

    fs::create_dir_all(dst)?;
    for entry in fs::read_dir(src)? {
        let entry = entry?;
        let src_path = entry.path();
        let dst_path = dst.join(entry.file_name());
        if src_path.is_dir() {
            copy_tree_if_missing(&src_path, &dst_path)?;
        } else if !dst_path.exists() {
            fs::copy(src_path, dst_path)?;
        }
    }
    Ok(())
}

#[cfg(test)]
mod tests {
    use super::*;
    use tempfile::TempDir;

    #[test]
    fn test_is_protected_path() {
        assert!(is_protected_path("SOUL.md"));
        assert!(is_protected_path("workspace/SOUL.md"));
        assert!(is_protected_path("/workspace/USER.md"));
        assert!(is_protected_path(r"workspace\AGENTS.md"));
        assert!(is_protected_path("TOOLS.md"));
        assert!(!is_protected_path("notes.md"));
    }

    #[test]
    fn test_queue_and_apply_refreshes_snapshot() {
        let tmp = TempDir::new().unwrap();
        let char_dir = tmp.path().join("data").join("TestChar");
        let config_dir = tmp.path().join("config");
        let workspace = character_workspace_dir(&config_dir, "TestChar");

        fs::create_dir_all(&workspace).unwrap();
        fs::write(workspace.join(SOUL_FILE), "new soul").unwrap();
        fs::write(workspace.join(USER_FILE), "new user").unwrap();
        fs::write(workspace.join(AGENTS_FILE), "new agents").unwrap();
        fs::write(workspace.join(TOOLS_FILE), "new tools").unwrap();
        fs::write(workspace.join(HEARTBEAT_FILE), "new heartbeat").unwrap();

        queue_deferred_edit(&char_dir, "workspace/SOUL.md").unwrap();
        queue_deferred_edit(&char_dir, "USER.md").unwrap();

        apply_deferred_edits(&char_dir, &config_dir, "TestChar").unwrap();

        assert_eq!(
            fs::read_to_string(active_prompt_file(&char_dir, SOUL_FILE)).unwrap(),
            "new soul"
        );
        assert_eq!(
            fs::read_to_string(active_prompt_file(&char_dir, USER_FILE)).unwrap(),
            "new user"
        );
        assert!(!char_dir.join("deferred_edits.jsonl").exists());
    }

    #[test]
    fn test_pending_paths_deduped() {
        let tmp = TempDir::new().unwrap();
        let char_dir = tmp.path().join("char");
        fs::create_dir_all(&char_dir).unwrap();

        queue_deferred_edit(&char_dir, "workspace/SOUL.md").unwrap();
        queue_deferred_edit(&char_dir, "SOUL.md").unwrap();
        queue_deferred_edit(&char_dir, "AGENTS.md").unwrap();

        let paths = pending_deferred_edit_paths(&char_dir).unwrap();
        assert_eq!(paths, vec!["AGENTS.md".to_string(), "SOUL.md".to_string()]);
    }

    #[test]
    fn test_workspace_migration_and_global_seed() {
        let tmp = TempDir::new().unwrap();
        let char_dir = tmp.path().join("data").join("TestChar");
        let config_dir = tmp.path().join("config");
        let char_config = character_config_dir(&config_dir, "TestChar");

        fs::create_dir_all(char_config.join("prompts")).unwrap();
        fs::write(char_config.join("character.md"), "orig soul").unwrap();
        fs::write(char_config.join("user.md"), "orig user").unwrap();
        fs::write(char_config.join("prompts/system.md"), "orig agents").unwrap();
        fs::write(config_dir.join("user.md"), "global user").unwrap();
        fs::create_dir_all(char_dir.join("memories").join("daily")).unwrap();
        fs::write(
            char_dir
                .join("memories")
                .join("daily")
                .join("2026-01-01.md"),
            "note",
        )
        .unwrap();

        ensure_character_workspace(&char_dir, &config_dir, "TestChar").unwrap();

        let workspace = character_workspace_dir(&config_dir, "TestChar");
        assert_eq!(
            fs::read_to_string(workspace.join(SOUL_FILE)).unwrap(),
            "orig soul"
        );
        assert_eq!(
            fs::read_to_string(workspace.join(USER_FILE)).unwrap(),
            "orig user"
        );
        assert_eq!(
            fs::read_to_string(workspace.join(AGENTS_FILE)).unwrap(),
            "orig agents"
        );
        assert!(workspace.join(TOOLS_FILE).exists());
        assert!(workspace.join(HEARTBEAT_FILE).exists());
        assert_eq!(
            fs::read_to_string(
                character_memory_dir(&config_dir, "TestChar").join("daily/2026-01-01.md")
            )
            .unwrap(),
            "note"
        );
    }

    #[test]
    fn test_snapshot_seed_only_when_missing() {
        let tmp = TempDir::new().unwrap();
        let char_dir = tmp.path().join("data").join("TestChar");
        let config_dir = tmp.path().join("config");
        let workspace = character_workspace_dir(&config_dir, "TestChar");

        fs::create_dir_all(&workspace).unwrap();
        fs::write(workspace.join(SOUL_FILE), "workspace soul").unwrap();
        fs::write(workspace.join(USER_FILE), "workspace user").unwrap();
        fs::write(workspace.join(AGENTS_FILE), "workspace agents").unwrap();
        fs::write(workspace.join(TOOLS_FILE), "workspace tools").unwrap();
        fs::write(workspace.join(HEARTBEAT_FILE), "workspace heartbeat").unwrap();

        ensure_active_prompt_snapshot(&char_dir, &config_dir, "TestChar").unwrap();
        fs::write(workspace.join(SOUL_FILE), "edited later").unwrap();
        ensure_active_prompt_snapshot(&char_dir, &config_dir, "TestChar").unwrap();

        assert_eq!(
            fs::read_to_string(active_prompt_file(&char_dir, SOUL_FILE)).unwrap(),
            "workspace soul"
        );
    }

    #[test]
    fn test_protected_edit_stays_deferred_until_apply() {
        let tmp = TempDir::new().unwrap();
        let char_dir = tmp.path().join("data").join("TestChar");
        let config_dir = tmp.path().join("config");
        let workspace = character_workspace_dir(&config_dir, "TestChar");

        fs::create_dir_all(&workspace).unwrap();
        fs::write(workspace.join(SOUL_FILE), "active soul").unwrap();
        fs::write(workspace.join(USER_FILE), "active user").unwrap();
        fs::write(workspace.join(AGENTS_FILE), "active agents").unwrap();
        fs::write(workspace.join(TOOLS_FILE), "active tools").unwrap();
        fs::write(workspace.join(HEARTBEAT_FILE), "active heartbeat").unwrap();

        ensure_active_prompt_snapshot(&char_dir, &config_dir, "TestChar").unwrap();

        fs::write(workspace.join(SOUL_FILE), "edited soul").unwrap();
        queue_deferred_edit(&char_dir, "SOUL.md").unwrap();
        ensure_active_prompt_snapshot(&char_dir, &config_dir, "TestChar").unwrap();

        assert_eq!(
            fs::read_to_string(active_prompt_file(&char_dir, SOUL_FILE)).unwrap(),
            "active soul",
            "protected workspace edits must not live-activate before apply"
        );
        assert_eq!(
            pending_deferred_edit_paths(&char_dir).unwrap(),
            vec!["SOUL.md"]
        );

        apply_deferred_edits(&char_dir, &config_dir, "TestChar").unwrap();

        assert_eq!(
            fs::read_to_string(active_prompt_file(&char_dir, SOUL_FILE)).unwrap(),
            "edited soul",
            "apply_deferred_edits is the activation boundary"
        );
        assert!(!char_dir.join("deferred_edits.jsonl").exists());
    }
}

File: backend/daemon-ts/src/memory/deferred_edits.ts
/**
 * Deferred edits + active prompt snapshot.
 *
 * Port of `backend/daemon/src/memory/deferred_edits.rs`.
 *
 * Edits to "prompt-visible" workspace files (SOUL.md, USER.md, AGENTS.md,
 * TOOLS.md, HEARTBEAT.md, MEMORY.md) are visible on the filesystem
 * immediately but only become *prompt-active* after the next compaction
 * boundary. The snapshot under `<characterDataDir>/active_prompt/` is the
 * version the system prompt reads from; `apply_deferred_edits` refreshes
 * the snapshot and clears the queue.
 *
 * `tools/paths.ts` already owns the normalization helpers
 * (`normalizeProtectedPath`, `normalizePromptVisiblePath`, etc.) — we
 * re-export them here so the conceptual home (`memory/`) matches the
 * Rust layout without duplicating the implementation.
 */

import fs from "node:fs";
import path from "node:path";

import {
  isPromptVisiblePath,
  isProtectedPath,
  normalizePromptVisiblePath,
  normalizeProtectedPath,
} from "../tools/paths.ts";

// ---------------------------------------------------------------------------
// Re-exports — single source of truth lives in tools/paths.ts
// ---------------------------------------------------------------------------

export {
  isPromptVisiblePath,
  isProtectedPath,
  normalizePromptVisiblePath,
  normalizeProtectedPath,
};

// ---------------------------------------------------------------------------
// Constants
// ---------------------------------------------------------------------------

export const SOUL_FILE = "SOUL.md";
export const USER_FILE = "USER.md";
export const AGENTS_FILE = "AGENTS.md";
export const TOOLS_FILE = "TOOLS.md";
export const HEARTBEAT_FILE = "HEARTBEAT.md";

/** Top-level prompt-visible workspace files (no MEMORY.md — that's the index). */
export const PROTECTED_PATHS = [
  SOUL_FILE,
  USER_FILE,
  AGENTS_FILE,
  TOOLS_FILE,
  HEARTBEAT_FILE,
] as const;

/** Active-prompt snapshot directory under the character data dir. */
const ACTIVE_PROMPT_DIR = "active_prompt";

export const MEMORY_INDEX_FILE = "MEMORY.md";
export const MEMORY_INDEX_DEFERRED_PATH = "MEMORY.md";

const LEGACY_RECENT_MEMORY_SNAPSHOT = "RECENT_MEMORY.md";

const DEFAULT_TOOLS_GUIDANCE = `# TOOLS

Use tools when they materially help.

- Read files before editing them.
- Search memory files before guessing facts about the user or past events.
- Prefer concise, direct tool use over busywork.
`;

const DEFAULT_HEARTBEAT_GUIDANCE = `# HEARTBEAT

- Use this private turn however seems useful.
- You may use tools, schedule the next wake, or send the user a message.
- If nothing needs action, respond HEARTBEAT_OK.
`;

const DEFERRED_EDITS_QUEUE = "deferred_edits.jsonl";

// ---------------------------------------------------------------------------
// Path helpers — inlined `shore_config::character_*` for self-containment
// ---------------------------------------------------------------------------

/** `<configDir>/characters/<name>/`. */
export function characterConfigDir(configDir: string, charName: string): string {
  return path.join(configDir, "characters", charName);
}

/** `<configDir>/characters/<name>/workspace/`. */
export function characterWorkspaceDir(configDir: string, charName: string): string {
  return path.join(characterConfigDir(configDir, charName), "workspace");
}

/** `<configDir>/characters/<name>/workspace/<file>`. */
export function characterWorkspaceFile(
  configDir: string,
  charName: string,
  file: string,
): string {
  return path.join(characterWorkspaceDir(configDir, charName), file);
}

/** `<configDir>/characters/<name>/workspace/memory/`. */
export function characterMemoryDir(configDir: string, charName: string): string {
  return path.join(characterWorkspaceDir(configDir, charName), "memory");
}

// ---------------------------------------------------------------------------
// Active prompt snapshot
// ---------------------------------------------------------------------------

export function activePromptDir(characterDataDir: string): string {
  return path.join(characterDataDir, ACTIVE_PROMPT_DIR);
}

export function activePromptFile(characterDataDir: string, name: string): string {
  return path.join(activePromptDir(characterDataDir), name);
}

export function memoryIndexPath(configDir: string, charName: string): string {
  return path.join(characterWorkspaceDir(configDir, charName), MEMORY_INDEX_FILE);
}

/**
 * Load the prompt-visible memory index. If the active snapshot file
 * exists, that's the answer (even if empty → `undefined`) — an empty
 * snapshot is the sentinel that says "canonical edit is pending
 * compaction, don't surface it yet." Falls back to canonical only when
 * the snapshot doesn't exist at all.
 */
export function loadMemoryIndex(
  characterDataDir: string,
  configDir: string,
  charName: string,
): string | undefined {
  const active = activePromptFile(characterDataDir, MEMORY_INDEX_FILE);
  if (fs.existsSync(active)) return readNonEmpty(active);
  return readNonEmpty(memoryIndexPath(configDir, charName));
}

export function loadCanonicalMemoryIndex(
  configDir: string,
  charName: string,
): string | undefined {
  return readNonEmpty(memoryIndexPath(configDir, charName));
}

export function loadActivePromptFile(
  characterDataDir: string,
  name: string,
): string | undefined {
  return readNonEmpty(activePromptFile(characterDataDir, name));
}

function readNonEmpty(file: string): string | undefined {
  try {
    const content = fs.readFileSync(file, "utf8");
    if (content.trim().length === 0) return undefined;
    return content;
  } catch (e) {
    if ((e as NodeJS.ErrnoException).code === "ENOENT") return undefined;
    throw e;
  }
}

// ---------------------------------------------------------------------------
// Deferred edit queue
// ---------------------------------------------------------------------------

/**
 * Append a queue entry for a prompt-visible workspace file. No-op on
 * non-prompt-visible paths so callers can pass through any edit path.
 *
 * Creates the queue file and, for MEMORY.md, a zero-byte sentinel under
 * the active-prompt directory so subsequent `ensure_active_prompt_snapshot`
 * calls don't re-seed it from canonical (the canonical version is now
 * pending activation).
 */
export function queueDeferredEdit(characterDataDir: string, p: string): void {
  const normalized = normalizePromptVisiblePath(p);
  if (normalized === undefined) return;

  fs.mkdirSync(characterDataDir, { recursive: true });
  if (normalized === MEMORY_INDEX_DEFERRED_PATH) {
    ensureDeferredMemoryIndexSentinel(characterDataDir);
  }

  const queuePath = path.join(characterDataDir, DEFERRED_EDITS_QUEUE);
  const line =
    JSON.stringify({
      path: normalized,
      timestamp: nowRfc3339Local(),
    }) + "\n";
  fs.appendFileSync(queuePath, line, "utf8");
}

/**
 * Wrapper for the common case of queueing a MEMORY.md refresh — keeps
 * compaction sites parallel to the Rust API.
 */
export function noteMemoryIndexDeferred(characterDataDir: string): void {
  queueDeferredEdit(characterDataDir, MEMORY_INDEX_DEFERRED_PATH);
}

/**
 * Return the deduped set of prompt-visible paths waiting for activation.
 * Malformed queue lines are skipped silently — mirrors Rust's tolerant
 * `serde_json::from_str` loop.
 */
export function pendingDeferredEditPaths(characterDataDir: string): string[] {
  const queuePath = path.join(characterDataDir, DEFERRED_EDITS_QUEUE);
  let content: string;
  try {
    content = fs.readFileSync(queuePath, "utf8");
  } catch (e) {
    if ((e as NodeJS.ErrnoException).code === "ENOENT") return [];
    throw e;
  }

  const seen = new Set<string>();
  for (const line of content.split("\n")) {
    const trimmed = line.trim();
    if (trimmed.length === 0) continue;
    let parsed: unknown;
    try {
      parsed = JSON.parse(trimmed);
    } catch {
      continue;
    }
    const obj = parsed as Record<string, unknown>;
    const raw = obj["path"];
    if (typeof raw !== "string") continue;
    const normalized = normalizePromptVisiblePath(raw);
    if (normalized !== undefined) seen.add(normalized);
  }
  return [...seen].sort();
}

/**
 * Refresh every protected file in the active snapshot from the canonical
 * workspace, then clear the queue. This is the compaction-boundary
 * activation step.
 */
export function applyDeferredEdits(
  characterDataDir: string,
  configDir: string,
  charName: string,
): void {
  refreshActivePromptSnapshot(characterDataDir, configDir, charName);

  const queuePath = path.join(characterDataDir, DEFERRED_EDITS_QUEUE);
  if (fs.existsSync(queuePath)) {
    fs.rmSync(queuePath);
  }
}

// ---------------------------------------------------------------------------
// Workspace bootstrap + snapshot seeding
// ---------------------------------------------------------------------------

/**
 * Ensure the character workspace layout exists and migrate any legacy
 * bootstrap files / memory directories into it. Idempotent: existing
 * files are left alone.
 *
 * Migrations performed:
 *   - characters/<C>/character.md → workspace/SOUL.md
 *   - characters/<C>/user.md      → workspace/USER.md
 *   - characters/<C>/prompts/system.md → workspace/AGENTS.md
 *   - <configDir>/user.md         → workspace/USER.md (only if absent)
 *   - <dataDir>/<C>/memories/**   → workspace/memory/**
 *
 * Default TOOLS.md and HEARTBEAT.md are written if absent.
 */
export function ensureCharacterWorkspace(
  characterDataDir: string,
  configDir: string,
  charName: string,
): void {
  const charConfigDir = characterConfigDir(configDir, charName);
  const workspaceDir = characterWorkspaceDir(configDir, charName);
  const memoryDir = characterMemoryDir(configDir, charName);

  fs.mkdirSync(workspaceDir, { recursive: true });
  fs.mkdirSync(memoryDir, { recursive: true });

  migrateLegacyFile(
    path.join(charConfigDir, "character.md"),
    path.join(workspaceDir, SOUL_FILE),
  );
  migrateLegacyFile(
    path.join(charConfigDir, "user.md"),
    path.join(workspaceDir, USER_FILE),
  );
  migrateLegacyFile(
    path.join(charConfigDir, "prompts", "system.md"),
    path.join(workspaceDir, AGENTS_FILE),
  );

  const globalUser = path.join(configDir, "user.md");
  const workspaceUser = path.join(workspaceDir, USER_FILE);
  if (fs.existsSync(globalUser) && !fs.existsSync(workspaceUser)) {
    fs.copyFileSync(globalUser, workspaceUser);
  }

  writeDefaultIfMissing(path.join(workspaceDir, TOOLS_FILE), DEFAULT_TOOLS_GUIDANCE);
  writeDefaultIfMissing(
    path.join(workspaceDir, HEARTBEAT_FILE),
    DEFAULT_HEARTBEAT_GUIDANCE,
  );

  const legacyMemories = path.join(characterDataDir, "memories");
  if (fs.existsSync(legacyMemories)) {
    copyTreeIfMissing(legacyMemories, memoryDir);
  }
}

/**
 * Seed the active-prompt snapshot from the canonical workspace. Existing
 * snapshot files are left untouched (this is the gate that keeps edits
 * deferred). Also cleans up the legacy `RECENT_MEMORY.md` snapshot from
 * the pre-rename era.
 */
export function ensureActivePromptSnapshot(
  characterDataDir: string,
  configDir: string,
  charName: string,
): void {
  ensureCharacterWorkspace(characterDataDir, configDir, charName);

  const activeDir = activePromptDir(characterDataDir);
  fs.mkdirSync(activeDir, { recursive: true });

  for (const name of PROTECTED_PATHS) {
    copyPromptVisibleFile(characterDataDir, configDir, charName, name, true);
  }
  copyPromptVisibleFile(
    characterDataDir,
    configDir,
    charName,
    MEMORY_INDEX_DEFERRED_PATH,
    true,
  );

  const legacy = path.join(activeDir, LEGACY_RECENT_MEMORY_SNAPSHOT);
  if (fs.existsSync(legacy)) fs.rmSync(legacy);
}

/**
 * Refresh every protected file (and MEMORY.md) in the active snapshot
 * from the canonical workspace. Called at the compaction boundary by
 * `applyDeferredEdits`.
 */
export function refreshActivePromptSnapshot(
  characterDataDir: string,
  configDir: string,
  charName: string,
): void {
  ensureCharacterWorkspace(characterDataDir, configDir, charName);
  for (const name of PROTECTED_PATHS) {
    copyPromptVisibleFile(characterDataDir, configDir, charName, name, false);
  }
  copyPromptVisibleFile(
    characterDataDir,
    configDir,
    charName,
    MEMORY_INDEX_DEFERRED_PATH,
    false,
  );
}

// ---------------------------------------------------------------------------
// Internal — file ops
// ---------------------------------------------------------------------------

function canonicalPromptVisibleFile(
  configDir: string,
  charName: string,
  p: string,
): string {
  if (p === MEMORY_INDEX_DEFERRED_PATH) {
    return memoryIndexPath(configDir, charName);
  }
  return characterWorkspaceFile(configDir, charName, p);
}

function activePromptSnapshotName(p: string): string {
  return p === MEMORY_INDEX_DEFERRED_PATH ? MEMORY_INDEX_FILE : p;
}

/**
 * Copy a canonical prompt-visible file into the active snapshot. With
 * `seedOnly`, an existing snapshot file is left intact (this is what
 * keeps deferred edits from leaking into the prompt). Without it
 * (refresh mode), the snapshot is overwritten; if the canonical file
 * has been deleted, the snapshot is removed too.
 */
function copyPromptVisibleFile(
  characterDataDir: string,
  configDir: string,
  charName: string,
  p: string,
  seedOnly: boolean,
): void {
  const activeDir = activePromptDir(characterDataDir);
  fs.mkdirSync(activeDir, { recursive: true });

  const src = canonicalPromptVisibleFile(configDir, charName, p);
  const dst = path.join(activeDir, activePromptSnapshotName(p));

  if (seedOnly && fs.existsSync(dst)) return;

  if (fs.existsSync(src)) {
    fs.copyFileSync(src, dst);
  } else if (!seedOnly && fs.existsSync(dst)) {
    fs.rmSync(dst);
  }
}

/**
 * Drop a zero-byte sentinel at active_prompt/MEMORY.md so subsequent
 * `ensureActivePromptSnapshot` calls (which run in seed-only mode) won't
 * pull the unactivated canonical version into the snapshot. Idempotent.
 */
function ensureDeferredMemoryIndexSentinel(characterDataDir: string): void {
  const dst = activePromptFile(characterDataDir, MEMORY_INDEX_FILE);
  if (fs.existsSync(dst)) return;
  fs.mkdirSync(path.dirname(dst), { recursive: true });
  fs.writeFileSync(dst, "");
}

function migrateLegacyFile(src: string, dst: string): void {
  if (fs.existsSync(src) && !fs.existsSync(dst)) {
    fs.mkdirSync(path.dirname(dst), { recursive: true });
    fs.copyFileSync(src, dst);
  }
}

function writeDefaultIfMissing(file: string, content: string): void {
  if (fs.existsSync(file)) return;
  fs.mkdirSync(path.dirname(file), { recursive: true });
  fs.writeFileSync(file, content);
}

function copyTreeIfMissing(src: string, dst: string): void {
  const stat = fs.statSync(src);
  if (stat.isFile()) {
    if (!fs.existsSync(dst)) {
      fs.mkdirSync(path.dirname(dst), { recursive: true });
      fs.copyFileSync(src, dst);
    }
    return;
  }
  fs.mkdirSync(dst, { recursive: true });
  for (const entry of fs.readdirSync(src)) {
    const childSrc = path.join(src, entry);
    const childDst = path.join(dst, entry);
    if (fs.statSync(childSrc).isDirectory()) {
      copyTreeIfMissing(childSrc, childDst);
    } else if (!fs.existsSync(childDst)) {
      fs.copyFileSync(childSrc, childDst);
    }
  }
}

function nowRfc3339Local(): string {
  const d = new Date();
  const pad = (n: number, w = 2): string => String(n).padStart(w, "0");
  const tzOffsetMin = -d.getTimezoneOffset();
  const sign = tzOffsetMin >= 0 ? "+" : "-";
  const absMin = Math.abs(tzOffsetMin);
  const offHours = Math.floor(absMin / 60);
  const offMins = absMin % 60;
  const offsetStr = `${sign}${pad(offHours)}:${pad(offMins)}`;
  return (
    `${d.getFullYear()}-${pad(d.getMonth() + 1)}-${pad(d.getDate())}T` +
    `${pad(d.getHours())}:${pad(d.getMinutes())}:${pad(d.getSeconds())}` +
    offsetStr
  );
}
